internal/authorizer/authorize_test.go

```diff
@@ -0,0 +1,277 @@
+	}
+	namespaceResponse := types.DataResponseV1{
+		Result: &namespaceResult,
+	}
+
+	var namespaceResultDeny interface{} = map[string]string{
+		"allowed": "false",
+		"data":    `{"matchers":[{"Type":2,"Name":"kubernetes_namespace_name","Value":""}],"matcherOp":"or"}`,
+	}
+	namespaceResponseDeny := types.DataResponseV1{
+		Result: &namespaceResultDeny,
+	}
+	tt := []struct {
+		desc          string
+		matcher       *config.Matcher
+		cacheResponse types.DataResponseV1
+		cacheFound    bool
+		cacheGetErr   error
+		cacheSetErr   error
+		sarFunc       sarFunc
+		nsList        []string
+		nsErr         error
+		verb          string
+		namespaces    []string
+		path          string
+		wantAuthorize types.DataResponseV1
+		wantErr       error
+	}{
+		{
+			desc:    "allow - get, no matcher",
+			matcher: config.EmptyMatcher(),
+			sarFunc: allowSAR,
+			nsList: []string{
+				"test-namespace-1",
+			},
+			verb:          GetVerb,
+			wantAuthorize: minimalDataResponseV1(true),
+			wantErr:       nil,
+		},
+		{
+			desc:    "allow - get, with matcher",
+			matcher: namespaceMatcher,
+			sarFunc: allowSAR,
+			nsList: []string{
+				"test-namespace-1",
+			},
+			verb:          GetVerb,
+			wantAuthorize: namespaceResponse,
+			wantErr:       nil,
+		},
+		{
+			desc:          "deny - get, with matcher, no namespaces",
+			matcher:       namespaceMatcher,
+			sarFunc:       allowSAR,
+			nsList:        []string{},
+			verb:          GetVerb,
+			wantAuthorize: namespaceResponseDeny,
+			wantErr:       nil,
+		},
+		{
+			desc:          "allow - create",
+			matcher:       config.EmptyMatcher(),
+			sarFunc:       allowSAR,
+			nsList:        []string{},
+			verb:          CreateVerb,
+			wantAuthorize: minimalDataResponseV1(true),
+			wantErr:       nil,
+		},
+		{
+			desc:        "fail - cache get error",
+			cacheGetErr: errors.New("get-cache error"),
+			verb:        GetVerb,
+			wantErr:     errors.New("failed to fetch authorization response from cache: get-cache error"),
+		},
+		{
+			desc:    "fail - wrong verb",
+			matcher: config.EmptyMatcher(),
+			sarFunc: allowSAR,
+			nsList:  []string{},
+			verb:    "invalid",
+			wantErr: errors.New("unexpected verb: invalid"),
+		},
+		{
+			desc:    "fail - SAR error",
+			matcher: config.EmptyMatcher(),
+			sarFunc: simpleSARFunc(false, errors.New("test SAR error")),
+			nsList: []string{
+				"test-namespace-1",
+			},
+			verb:    GetVerb,
+			wantErr: errors.New("cluster-wide SAR failed: test SAR error"),
+		},
+		{
+			desc:    "fail - list namespace error",
+			matcher: namespaceMatcher,
+			sarFunc: allowSAR,
+			nsErr:   errors.New("test list namespace error"),
+			verb:    GetVerb,
+			wantErr: errors.New("failed to access api server: test list namespace error"),
+		},
+		{
+			desc:          "allow - cached",
+			matcher:       config.EmptyMatcher(),
+			cacheResponse: namespaceResponse,
+			cacheFound:    true,
+			verb:          GetVerb,
+			wantAuthorize: namespaceResponse,
+		},
```

Comment:
Out of curiousity: The request matcher is empty but the cached one is a regex matcher. Is this case something we should omit from happening? AFAIU the matchers are CLI options which can only change if the container spec changes during upgrades. I can not figure out any case that breaks here, but maybe it's worth to call this out even in the case of no issue. WDYT?

Reply:
I think I just forgot to remove the unused line in this test case. It does not get to the point of checking the matcher. The matcher depends on the `tenant` and `groups`, so I don't think there's a risk of a cache hit that returns the wrong matcher.